File: pkg/tmuxctl/tmux.go
// Package tmuxctl is a thin wrapper around the tmux CLI for the operations
// allagent needs: spawn a detached session running a backend, send keys to it,
// capture its visible pane, resize it, and kill it.
package tmuxctl

import (
	"bytes"
	"errors"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
)

// Session represents a detached tmux session hosting a single backend pane.
type Session struct {
	Name    string
	Command []string
}

// HasSession reports whether a tmux session with the given name exists.
func HasSession(name string) bool {
	cmd := exec.Command("tmux", "has-session", "-t", name)
	return cmd.Run() == nil
}

// NewSession creates a detached tmux session running `command` inside a pane
// of the requested size. If a session with the same name already exists it is
// killed first.
func NewSession(name string, width, height int, workingDir string, command []string) (*Session, error) {
	if len(command) == 0 {
		return nil, errors.New("tmuxctl: command is required")
	}
	if HasSession(name) {
		if err := KillSession(name); err != nil {
			return nil, fmt.Errorf("kill existing session: %w", err)
		}
	}
	args := []string{
		"new-session", "-d",
		"-s", name,
		"-x", strconv.Itoa(width),
		"-y", strconv.Itoa(height),
	}
	if workingDir != "" {
		args = append(args, "-c", workingDir)
	}
	args = append(args, command...)

	cmd := exec.Command("tmux", args...)
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("tmux new-session: %w (stderr=%s)", err, stderr.String())
	}
	_ = exec.Command("tmux", "set-option", "-t", name, "status", "off").Run()
	return &Session{Name: name, Command: command}, nil
}

// KillSession terminates a tmux session by name (no-op if absent).
func KillSession(name string) error {
	if !HasSession(name) {
		return nil
	}
	return exec.Command("tmux", "kill-session", "-t", name).Run()
}

func (s *Session) Kill() error    { return KillSession(s.Name) }
func (s *Session) Alive() bool    { return HasSession(s.Name) }
func (s *Session) target() string { return s.Name + ":0.0" }

// Resize changes the pane size of the session's only window.
func (s *Session) Resize(width, height int) error {
	return exec.Command("tmux", "resize-window",
		"-t", s.Name,
		"-x", strconv.Itoa(width),
		"-y", strconv.Itoa(height),
	).Run()
}

// SendKeys delivers a literal string to the pane (no Enter).
func (s *Session) SendKeys(input string) error {
	if input == "" {
		return nil
	}
	return exec.Command("tmux", "send-keys", "-t", s.target(), "-l", "--", input).Run()
}

// SendLine sends `input` followed by Enter.
func (s *Session) SendLine(input string) error {
	if input != "" {
		if err := s.SendKeys(input); err != nil {
			return err
		}
	}
	return s.SendSpecial("Enter")
}

// SendSpecial sends named keys like "Enter", "C-c", "Up", "Escape", etc.
func (s *Session) SendSpecial(keys ...string) error {
	if len(keys) == 0 {
		return nil
	}
	args := append([]string{"send-keys", "-t", s.target()}, keys...)
	return exec.Command("tmux", args...).Run()
}

// Interrupt sends Ctrl+C.
func (s *Session) Interrupt() error { return s.SendSpecial("C-c") }

// CaptureOpts controls CapturePane.
type CaptureOpts struct {
	WithEscapes       bool
	IncludeScrollback bool
}

// CapturePane returns the current contents of the backend pane.
func (s *Session) CapturePane(opts CaptureOpts) (string, error) {
	args := []string{"capture-pane", "-p", "-t", s.target()}
	if opts.WithEscapes {
		args = append(args, "-e")
	}
	if opts.IncludeScrollback {
		args = append(args, "-S", "-")
	}
	args = append(args, "-J")

	cmd := exec.Command("tmux", args...)
	var out, stderr bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("tmux capture-pane: %w (stderr=%s)", err, stderr.String())
	}
	return out.String(), nil
}

// PanePID returns the shell PID of the session's pane.
func (s *Session) PanePID() (int, error) {
	cmd := exec.Command("tmux", "display-message", "-p", "-t", s.target(), "#{pane_pid}")
	var out bytes.Buffer
	cmd.Stdout = &out
	if err := cmd.Run(); err != nil {
		return 0, err
	}
	return strconv.Atoi(strings.TrimSpace(out.String()))
}

File: packages/claudraband-core/src/tmuxctl/tmux.ts
import { spawn, spawnSync } from "child_process";
import { awaitPaneIdle } from "../terminal/activity";
import type { PaneActivityOptions, ActivityResult } from "../terminal/activity";

export interface TmuxWindowSummary {
  windowId: string;
  paneId: string;
  windowName: string;
  paneCurrentPath?: string;
  windowActivity?: string;
  panePid?: number;
}

async function tmux(...args: string[]): Promise<{ stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    const proc = spawn("bash", ["-lc", shellCommand("tmux", ...args)]);
    const out: string[] = [];
    const err: string[] = [];
    proc.stdout.on("data", (d: Buffer) => out.push(d.toString()));
    proc.stderr.on("data", (d: Buffer) => err.push(d.toString()));
    proc.on("close", (code) => {
      if (code === 0) {
        resolve({ stdout: out.join(""), stderr: err.join("") });
      } else {
        reject(
          new Error(
            `tmux ${args.join(" ")}: exit ${code} (stderr=${err.join("")})`,
          ),
        );
      }
    });
    proc.on("error", reject);
  });
}

function shellCommand(...args: string[]): string {
  return args.map(shellQuote).join(" ");
}

function shellQuote(arg: string): string {
  return `'${arg.replace(/'/g, "'\\''")}'`;
}

export function hasSession(name: string): boolean {
  return (
    spawnSync("bash", ["-lc", shellCommand("tmux", "has-session", "-t", name)], {
      stdio: "pipe",
    }).status === 0
  );
}

function hasPane(id: string): boolean {
  return (
    spawnSync(
      "bash",
      ["-lc", shellCommand("tmux", "display-message", "-p", "-t", id, "#{pane_id}")],
      { stdio: "pipe" },
    ).status === 0
  );
}

function windowTarget(sessionName: string, windowName: string): string {
  return `${sessionName}:${windowName}`;
}

export async function killSession(name: string): Promise<void> {
  if (!hasSession(name)) return;
  await tmux("kill-session", "-t", name);
}

export async function listWindows(name: string): Promise<TmuxWindowSummary[]> {
  if (!hasSession(name)) return [];

  const result = await tmux(
    "list-windows",
    "-t",
    name,
    "-F",
    "#{window_id}\t#{pane_id}\t#{window_name}\t#{pane_current_path}\t#{window_activity}\t#{pane_pid}",
  );

  return result.stdout
    .trim()
    .split("\n")
    .filter(Boolean)
    .map((line) => {
      const [
        windowId,
        paneId,
        windowName,
        paneCurrentPath,
        windowActivity,
        panePid,
      ] = line.split("\t", 6);
      return {
        windowId,
        paneId,
        windowName,
        paneCurrentPath: paneCurrentPath || undefined,
        windowActivity: windowActivity || undefined,
        panePid: panePid ? parseInt(panePid, 10) : undefined,
      };
    })
    .filter((window) => window.windowId && window.paneId && window.windowName);
}

export interface CaptureOpts {
  withEscapes?: boolean;
  includeScrollback?: boolean;
}

export class Session {
  readonly name: string;
  readonly command: string[];
  readonly windowId: string;
  readonly paneId: string;

  private constructor(
    name: string,
    command: string[],
    windowId: string,
    paneId: string,
  ) {
    this.name = name;
    this.command = command;
    this.windowId = windowId;
    this.paneId = paneId;
  }

  static async newSession(
    name: string,
    width: number,
    height: number,
    workingDir: string,
    command: string[],
    windowName: string,
  ): Promise<Session> {
    if (command.length === 0) {
      throw new Error("tmuxctl: command is required");
    }

    const format = "#{window_id}\t#{pane_id}";
    const target = windowTarget(name, windowName);
    const result = hasSession(name)
      ? await tmux(
        "new-window",
        "-P",
        "-F",
        format,
        "-t",
        name,
        "-n",
        windowName,
        ...(workingDir ? ["-c", workingDir] : []),
        ...command,
        ";",
        "set-option",
        "-q",
        "-t",
        name,
        "destroy-unattached",
        "off",
        ";",
        "set-option",
        "-q",
        "-t",
        name,
        "status",
        "off",
        ";",
        "resize-window",
        "-t",
        target,
        "-x",
        String(width),
        "-y",
        String(height),
      )
      : await tmux(
        "new-session",
        "-d",
        "-P",
        "-F",
        format,
        "-s",
        name,
        "-n",
        windowName,
        "-x",
        String(width),
        "-y",
        String(height),
        ...(workingDir ? ["-c", workingDir] : []),
        ...command,
        ";",
        "set-option",
        "-q",
        "-t",
        name,
        "destroy-unattached",
        "off",
        ";",
        "set-option",
        "-q",
        "-t",
        name,
        "status",
        "off",
        ";",
        "resize-window",
        "-t",
        target,
        "-x",
        String(width),
        "-y",
        String(height),
      );

    const [windowId, paneId] = result.stdout.trim().split(/\s+/, 2);
    if (!windowId || !paneId) {
      throw new Error(`tmuxctl: failed to parse tmux target ids: ${result.stdout}`);
    }

    return new Session(name, command, windowId, paneId);
  }

  /**
   * Find an existing tmux window by session name and window name.
   * Returns a Session if the window is alive, null otherwise.
   */
  static async find(
    sessionName: string,
    windowName: string,
  ): Promise<Session | null> {
    if (!hasSession(sessionName)) return null;
    try {
      const result = await tmux(
        "list-windows",
        "-t",
        sessionName,
        "-F",
        "#{window_id}\t#{pane_id}\t#{window_name}",
      );
      for (const line of result.stdout.trim().split("\n")) {
        const [windowId, paneId, name] = line.split("\t", 3);
        if (name === windowName && windowId && paneId) {
          const session = new Session(sessionName, [], windowId, paneId);
          if (session.isAlive) return session;
        }
      }
    } catch {
      // Session might not exist or list-windows failed.
    }
    return null;
  }

  async kill(): Promise<void> {
    if (!this.isAlive) return;
    await tmux("kill-window", "-t", this.windowId);
  }

  get isAlive(): boolean {
    return hasPane(this.paneId);
  }

  async resize(width: number, height: number): Promise<void> {
    await tmux(
      "resize-window",
      "-t",
      this.windowId,
      "-x",
      String(width),
      "-y",
      String(height),
    );
  }

  async sendKeys(input: string): Promise<void> {
    if (!input) return;
    await tmux("send-keys", "-t", this.paneId, "-l", "--", input);
  }

  async sendLine(input: string): Promise<void> {
    if (input) {
      await this.sendKeys(input);
    }
    await this.sendSpecial("Enter");
  }

  async sendSpecial(...keys: string[]): Promise<void> {
    if (keys.length === 0) return;
    await tmux("send-keys", "-t", this.paneId, ...keys);
  }

  async interrupt(): Promise<void> {
    await this.sendSpecial("C-c");
  }

  async capturePane(opts: CaptureOpts = {}): Promise<string> {
    const args = ["capture-pane", "-p", "-t", this.paneId];
    if (opts.withEscapes) {
      args.push("-e");
    }
    if (opts.includeScrollback) {
      args.push("-S", "-");
    }
    args.push("-J");

    const result = await tmux(...args);
    return result.stdout;
  }

  async awaitIdle(options?: PaneActivityOptions): Promise<ActivityResult> {
    return awaitPaneIdle(() => this.capturePane(), options);
  }

  async panePID(): Promise<number> {
    const result = await tmux(
      "display-message",
      "-p",
      "-t",
      this.paneId,
      "#{pane_pid}",
    );
    return parseInt(result.stdout.trim(), 10);
  }
}
